components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,211 @@
+  if (resource_bundle.IsGzipped(IDR_BRAVE_FEDERATED_CONFIG)) {
+    data_resource =
+        resource_bundle.LoadDataResourceString(IDR_BRAVE_FEDERATED_CONFIG);
+  } else {
+    data_resource = static_cast<std::string>(
+        resource_bundle.GetRawDataResource(IDR_BRAVE_FEDERATED_CONFIG));
+  }
+  DCHECK_GT(data_resource.size(), 0U);
+
+  lsc_ = std::make_unique<LearningServiceConfig>(data_resource);
+  const net::BackoffEntry::Policy reconnect_policy = lsc_->GetReconnectPolicy();
+  const net::BackoffEntry::Policy request_task_policy =
+      lsc_->GetRequestTaskPolicy();
+  const net::BackoffEntry::Policy post_results_policy =
+      lsc_->GetPostResultsPolicy();
+  model_spec_ = std::make_unique<ModelSpec>(lsc_->GetModelSpec());
+
+  communication_adapter_ = std::make_unique<CommunicationAdapter>(
+      url_loader_factory_, reconnect_policy, request_task_policy);
+  post_results_policy_ =
+      std::make_unique<const net::BackoffEntry::Policy>(post_results_policy);
+  post_results_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(post_results_policy_.get());
+
+  const int init_federated_service_wait_time_in_seconds =
+      brave_federated::features::GetInitFederatedServiceWaitTimeInSeconds();
+  init_task_timer_ = std::make_unique<base::OneShotTimer>();
+  init_task_timer_->Start(
+      FROM_HERE, base::Seconds(init_federated_service_wait_time_in_seconds),
+      this, &LearningService::Init);
+}
+
+LearningService::~LearningService() {
+  if (participating_) {
+    StopParticipating();
+  }
+
+  eligibility_observation_.Reset();
+}
+
+void LearningService::Init() {
+  DCHECK(url_loader_factory_);
+  DCHECK(eligibility_service_);
+  DCHECK(init_task_timer_);
+
+  VLOG(1) << "Initializing federated learning service.";
+  eligibility_observation_.Observe(eligibility_service_);
+  if (eligibility_service_->IsEligible()) {
+    StartParticipating();
+  }
+
+  VLOG(1) << "Eligibility: " << eligibility_service_->IsEligible();
+}
+
+void LearningService::OnEligibilityChanged(bool is_eligible) {
+  if (is_eligible) {
+    StartParticipating();
+    VLOG(2) << "Eligibility changed, started participating.";
+  } else {
+    StopParticipating();
+    VLOG(2) << "Eligibility changed, stopped participating.";
+  }
+}
+
+void LearningService::StartParticipating() {
+  if (participating_) {
+    return;
+  }
+
+  participating_ = true;
+  GetTasks();
+}
+
+void LearningService::StopParticipating() {
+  if (!participating_) {
+    return;
+  }
+
+  participating_ = false;
+  reconnect_timer_.reset();
+}
+
+void LearningService::GetTasks() {
+  communication_adapter_->GetTasks(base::BindOnce(
+      &LearningService::HandleTasksOrReconnect, weak_factory_.GetWeakPtr()));
+}
+
+void LearningService::HandleTasksOrReconnect(TaskList tasks, int reconnect) {
+  if (tasks.empty()) {
+    reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    reconnect_timer_->Start(FROM_HERE, base::Seconds(reconnect), this,
```

Comment:
race condition as discussed above ^^^

Reply:
as discussed you could just destroy communication_adapter_ when you stop participating